### ISSUE TITLE
API enhencement

### ISSUE THREAD

**Issue by goandylok** [m4b/scroll#11]
There are some problems of the API design: 
- Why `ctx` defaults to `Endian`? I can't see how generic the type is, and I think it's meaningless to select any of its variants to be default when ctx is not necessary. The unit `()` may be a better choice.
- Why this crate delivers `Sleb128` and `Leb128`? I think we could split them into a new crate, with dependency on `scroll`.
- Trait `FromCtx` and `TryFromCtx` has a lot of code duplication in implementation.
- Trait `TryRefFrom` requires ctx implements Copy when `RefFrom` don't.
- Lack of `MeasureWith` semantic. 
- The name. To be honest, I was confused at the beginning when reading the document and tring to figure out the relationship between a lot of traits in this crate. Finally, I came to realized that `pread` was actually a wrapper of trait `TryFromCtx` after reading source code. Also, the name of trait `IntoCtx` and `FromCtx` is vague and misleading. 

So I suggest the following modifications:
1. Remove `DefaultCtx` and replace by `()`
2. Rename `TryFromCtx`, `TryIntoCtx` to `TryRead`, `TryWrite`, which seem more fit their purposes. (the main idea is to reduce priority of the concept `Context` because this library is mainly to `scroll` across data then Read and Write things, but not working around some context)
3. Remove trait `FromCtx` `IntoCtx`, and then use `TryRead` and `TryWrite` in `cread` by `unwrap()` or `expect()`. This will make it easier to implement these traits.
4. Unify traits `RefFrom` and `TryRefFrom` `TryRefInto` into `TryRead` and `TryWrite`. This is a very early though and maybe I will open an pull request latter on this. In brief, `impl TryRefFrom for [usize]` is equal to `impl TryFrom for &[usize]`.
5. Split `SizeWith` into two new traits `WriteSize` and `ReadSize`. `WriteSize` takes a reference to self(&self) while `ReadSize` takes a reference to slice.
6. Move `Endian` into `scroll::ctx`, I think that's the right place for it.
7. Move traits `TryRead`, `TryWrite`, `WriteSize`, `ReadSize` to root module. I've two reasons to do that. First, they are not actually contexts, when the `Endian` is. Second, `TryRead` is related to `Pread`, `Cread` family, and they should have the same priority. This is one of the reasons I get confused in the beginning.
8. Rename `TryOffsetWith` => `TryOffset`. There is no ambiguity like that between `pread` and `pread_with`. Also, the reason for removing the word `with` is same a (2), is to reduce the attention on `context`.
9. Remove const `LE`, `BE` when `Endian::LE` is more idiomatic.
10. Remove const `LEB128` because it's just const `NATIVE`
11. Move const `NATIVE` and `NETWORK` into `scroll::ctx`
12. Remove type definition `Leb128`, which is ambiguous and just an alias to `NATIVE`.

Looking forward to your opinion and I think I can help with these refactoring.

**Comment by goandylok** [comment]
Addition question: Are we really need `pread`, `pread_with`, `pread_slice` and `pread_unsafe` instead of a single `pread`? 
- We could remove `CTX: Default` bound then remove `pread` method and rename `pread_with` to `pread`
- If we eliminate `TryRefFrom` finally, we can also remove `pread_slice` method.
- Does `pread_unsafe` seem similar to `cread`?

**Comment by m4b** [comment]
1. [ ] :+1: agreed
2. [ ] needs discussion
3. [ ] depends on 2. the perf of cread compared to pread is pretty different; try `cargo bench`.
4. [ ] needs discussion
5. [ ] I don't understand this proposal
6. [ ] :+1: agreed
7. [ ] :+1: tentatively agreed; ctx originally started as scratchpad for ideas on generics and context based parsing in scroll, and so the traits helping with this ended up there, but it makes implementing the traits by clients annoying and confusing. It's also a good datapoint where your initial confusion was; i want to reduce this as much as possible, as i think it's one of the reasons people haven't adopted as much, also because the trait definitions are terrifying to look at :rofl: 
8. [ ] :+1:
9. [ ] :-1: these are used to much, and so frequently, I think it's ok for them to exist somewhat unnaturally in the top module (we can re-export them tho, so they show up in ctx module); the most common thing people do in byteorder is rename `BigEndian => BE` on import.
10. [ ] :+1: 
11. [ ] :+1:
12. [ ] :+1: can't even remember why it was there; i think it was when i was experimenting with Pread being a trait to be used generically, like Read (but this won't scale because of multiple contexts)

I agree with almost everything you've proposed (and thanks for writing this list up, especially some of your pain points when reading/understanding the crate).

So, the major issue is 2 and the related points.  Let me just say, I changed the names of these traits several times; i even asked on IRC what people thought was a good name for the idea, especially `Into`.

I personally like the `TryFromCtx` and `FromCtx`; they have a natural and exact analog with the std traits; their only difference is that they provide another argument, the ctx, that's it.  If rust had default arguments, _this_ is what the traits should have been, as I frequently want to convert from one thing to another, but with an additional _context_ in order to do so, it really is a shame :/

That being said, I'm not totally against the name change; especially `Into` (it never really felt right to me, and I do tentatively like your Read/Write proposal).  However, and this brings me to the second point, is that there are technical reasons to have the `TryFromCtx` and `FromCtx` split - `FromCtx` doesn't provide a lifetime, and currently I can't see a way to return, generically, values from `TryFromCtx` in `Lread` (which does bind the return to a lifetime, which enables zero-copy &str parsing):

```rust
    fn lread_with<N: FromCtx<Ctx> + SizeWith<Ctx, Units = usize>>(&mut self, ctx: Ctx) -> Result<N> {
        let mut scratch = [0u8; 256];
        let size = N::size_with(&ctx);
        let mut buf = &mut scratch[0..size];
        self.read_exact(&mut buf)?;
        Ok(N::from_ctx(buf, ctx))
    }
```
`FromCtx` is lifetime-less, and so consequently, is permitted to generically return any value out of the scratch buffer it reads out of.

I think I could change it to have a buffer passed in which has the bytes copied into it, then the value is read out of this mutably borrowed buffer and lifetime should be attached to that, but I think the API here is totally awesome - it's simple, easy to use, and nice (provided your object isn't greater than 256 bytes). and I want to keep it this way.

Other miscellanous coments, the reason for `pread_slice` is that sometimes you need to read a str but you're only given it's length, not a delimiter (it might not even have one!), so you can't just `pread` it. Also you could in principle read out, say `&[ProgramHeader]`.

I think `pread_unsafe` existed because I thought I needed it to implement `Gread` iirc; but I don't believe it's necessary any longer.  It's hard to say, it's really pushing the limits of generics, so when you start hacking on it, you'll think it'll work then you'll get enormous and incredibly cryptic errors, and its hard to know what went wrong exactly.

In fact the majority of the tests in lib are just various ways of using pread as a client, to make sure a refactor didn't break it's usage downstream, which very often I wouldn't see until i tried to actually use it.  E.g., a recent example of this was https://github.com/m4b/goblin/issues/22 which was partially my fault since I concentrated on getting the method to compile, and not checking the scope of the lifetime, but this was all caused because the scope could never have been what i wanted, because it wasn't implemented for `[u8]` ! :D

> We could remove CTX: Default bound and then remove pread method , rename pread_with to pread, requiring caller to provide context, even it's just a unit. This help a lot in generic usage. At least, we should move the Default trait bound into a new impl block

I'm not sure what you mean precisely there; I don't think I'll rename pread_with -> pread; it was already that way, and I switched it and haven't looked back since.  Most people just want to do `bytes.pread::<usize>(0)` and be done with it; they don't care if it's LE, BE, and I'm usually one of those people.  I think using the context should be explicit, and I eventually settled on `with`, because it's a semi-Rusty idiom, it does actually read the value _with_ the context you give it, _and_ it's short.

So as I said above, I'm in agreement on most things here, but for `TryFromCtx` being sole trait, dumping `FromCtx` and renaming, I'd like to move along conservatively; honestly, I also think it's more important to finish #10 and get `TryFromOffset` working without all the offset logic in it, before thinking about something major like removing `TryFromCtx`; i'm still not sure how possible it is, there are probably bumps along the way.  Scroll is a fickle creature :dancer:

**Comment by goandylok** [comment]
There are so many things, and fortunately we reach a consensus.  Cheers!! 🍺 🍺 Still, there are something need to discuss in detail .

I understand your concern about the removal of `FromCtx`; I didn't think about the performance issue before because I thought we just invited a  `unwrap` while double bound check is gone. And now I will see this refactor as an experiment in my own fork, examining its feasibility and performance.

> the reason for pread_slice is that sometimes you need to read a str but you're only given it's length, not a delimiter

IMO, both the size and delimiter are ctx. This two are of the same purpose -- to tell where a str ends, and  they are so specific to the type that we should encapsulate this kind of logic into trait `TryFromCtx`. 
```rust
buffer.pread_slice::<&str>::(0, 20);
// =>
buffer.pread::<&str>::(0, StrCtx::Slice(20));
```

What's more, I want to give types ability to figure out the length of itself, that's what I purposed at (5), the `ReadSize` and `WriteSize`. In my practice, very few types have the exact same size of memory to byte representation except the primitives. 

For example, assume I would like to communication with hardware, steaming maybe some arrays of data. There would be a communication standard provided by the vendor, to tell how the data flow start, how it ends, and especially the how it store length. There are so many stuffs make its size bigger or smaller(compression). What the most apparent example is Leb128.

`ReadSize` takes the reference of self and then figure out how many space it will take in byte representation. Likewise, `WriteSize` takes reference to byte slice, and then figure out how many space that has been taken in byte slice.

As `SizeWith` is simply a wrapper of `size_of::<T>::()` and only implements for primitives, we can turn to impl `ReadSize` and `WriteSize` separately for primitives by using `size_of::<T>::()`.

And to me, to be honest, the name `TryFromCtx` really feels not good.

**Comment by m4b** [comment]
> buffer.pread::<&str>::(0, StrCtx::Length(20));

Omg I love this idea :heart: ; having the Ctx for strs (or others) be an enum didn't even occur to me, so cool !!!

> I understand your concern about the removal of FromCtx; I didn't think about the performance issue before because I thought we just invited a unwrap while double bound check was gone. And now I will experiment this refactor in my own fork, examining its feasibility and performance.

The big issue you'll see is lifetimes; for this reason, I think we cannot remove `FromCtx` (or whatever it ends up  being named); I suggest these dual `From/Into` versions be used for the "simple" types, or basic `repr(C)` type structs, and have both `cread` and `lread` use this as these conversion trait in the method signatures.

> What's more, I want to give types ability to figure out the length of itself, that's what I purposed at (5), the ReadSize and WriteSize. In my practice, very few types have the exact same size of memory to byte representation except the primitives.

I agree I want something here; I experimented heavily initially with a couple versions, but the trouble you'll find is that either you know the size ahead of time, in which case you can safely check the offset and return an error otherwise; _or_ you can try to read the thing out, and then you can update the offset correctly even if it's a variable length type, like `Leb128`, etc -- _BUT_ at the risk of out of bounds :/

I definitely want something here, and I'm interested in seeing your idea in practice, but I'm not sure yet how exactly it'll work.

Another idea I've been thinking about is a "Mutable Context"; instead of taking the Ctx by value, this particular method takes the context as a mutable reference (similar to Gread and it's offset); then, after reading, the context is updated, and hence reusable for further, _very context sensitive_ reading.

This would be particularly useful for reading out variable length strengths of all different sizes from a stream, and for example, perhaps the offset is inside the context, which gets updated after each string is read.  Lots of interesting things to work on.

> And to me, to be honest, the name TryFromCtx really feels not good.

Ok, seems you don't like this name at all :laughing: For now, I think we can punt on the name change, as this technically should be find-replace change, and will only pollute diffs, if you're experimenting with the underlying reader impls, how they interact with generic traits, etc., I think it's best to leave this part alone for now, as it will be easier for me to review changes, what you're trying to accomplish, etc.

**Comment by m4b** [comment]
So I just realized a minor compromise is that I believe that just for `pread` we can make the `where Ctx: Default`  be a requirement, and remove it as a trait requirement; thus it will only allow pread to be used in situations where the Ctx has default impl'd.  so if a user wants `pread_with` to be required, then they can force this behavior statically by not implementing Default for their context, etc.

Anyway, just a thought; I think more and more changing the `TryFromCtx` context from `(I, Ctx)` -> `Ctx` and moving all offsetting logic into pread itself is actually really important; though it's going to be an awful refactor, and will also massively break goblin downstream, but that's ok, as I think it's the right thing to do before 1.0

**Comment by goandylok** [comment]
I think I was blocked by issue #10 because this code largely depends on trait signatures.

**Comment by goandylok** [comment]
I wrote a small example that shows how the new `TryFromCtx` works. Note that `TryRefFrom` was removed since `TryFromCtx` can be impl on references.

### Definetion
```rust
pub trait TryFromCtx<'a, Ctx = ()>
    where Self: 'a + Sized
{
    fn try_from_ctx(scroll: &'a [u8], ctx: &Ctx) -> Result<(Self, usize), ()>;
}

trait Pread<'a> {
    fn pread_with<T, Ctx>(self, offset: usize, ctx: &Ctx) -> Result<T, ()>
        where T: TryFromCtx<'a, Ctx>,
              Self: 'a;
}

trait Gread<'a> {
    fn gread_with<T, Ctx>(self, offset: &mut usize, ctx: &Ctx) -> Result<T, ()>
        where T: TryFromCtx<'a, Ctx>,
              Self: 'a;
}
```

### Implementation for &[u8]
```rust
impl<'a> Pread<'a> for &'a [u8] {
    fn pread_with<T, Ctx>(self: &'a [u8], offset: usize, ctx: &Ctx) -> Result<T, ()>
        where T: TryFromCtx<'a, Ctx>,
              Self: 'a
    {
        TryFromCtx::try_from_ctx(&self[offset..], ctx).map(|(t, _)| t)
    }
}

impl<'a> Gread<'a> for &'a [u8] {
    fn gread_with<T, Ctx>(self: &'a [u8], offset: &mut usize, ctx: &Ctx) -> Result<T, ()>
        where T: TryFromCtx<'a, Ctx>,
              Self: 'a
    {
        TryFromCtx::try_from_ctx(&self[*offset..], ctx).map(|(t, size)| {
                                                                *offset += size;
                                                                t
                                                            })
    }
}
```

### Implementation for `&str`
```rust
#[derive(Debug)]
pub enum StrCtx {
    Delimiter(u8),
    Length(usize),
}

impl<'a> TryFromCtx<'a, StrCtx> for &'a str {
    fn try_from_ctx(scroll: &'a [u8], ctx: &StrCtx) -> Result<(Self, usize), ()> {
        let len = match *ctx {
            StrCtx::Delimiter(delimiter) => scroll.iter().take_while(|c| **c != delimiter).count(),
            StrCtx::Length(len) => len,
        };

        if len > scroll.len() {
            return Err(());
        };

        str::from_utf8(&scroll[..len])
            .map(|s| (s, len))
            .map_err(|_| ())
    }
}
```

### Usage
```rust
#[test]
fn test_str() {
    let bytes: &[u8] = b"hello, world!\0some_other_things";
    let s: &str = bytes.pread_with(0, &StrCtx::Delimiter(0)).unwrap();
    assert_eq!(s, "hello, world!");

    let bytes: &[u8] = b"abcdefghijklmnopqrstuvwxyz";
    let s: &str = bytes.pread_with(0, &StrCtx::Length(15)).unwrap();
    assert_eq!(s, "abcdefghijklmno");
    assert!(bytes
                .pread_with::<&str, _>(0, &StrCtx::Length(27))
                .is_err());

    let bytes: &[u8] = b"hello, world!\0some_other_things";
    let mut offset = 0;
    let s: &str = bytes
        .gread_with(&mut offset, &StrCtx::Delimiter(0))
        .unwrap();
    assert_eq!(s, "hello, world!");
    assert_eq!(offset, 13);
}
```

**Comment by goandylok** [comment]
I tried to implement `ReadSize` and it seemed too complex for the types which do not have a obvious variable length, like `Leb128`. But it works good with common types. In the example above, `pread` strips down the slice by offset then pass to `TryFromCtx`. In this example, after `pread` stripes slice's head, `pread` will also stripe slice's tail by the length messure by `ReadSize`.

### Definetion
```rust
pub trait TryFromCtx<'a, Ctx = ()>
    where Self: 'a + Sized
{
    fn try_from_ctx(scroll: &'a [u8], ctx: &Ctx) -> Result<Self, ()>;
}

pub trait ReadSize<'a, Ctx = ()>
    where Self: 'a + Sized
{
    fn read_size(scroll: &'a [u8], ctx: &Ctx) -> usize;
}

trait Pread<'a> {
    fn pread_with<T, Ctx>(self, offset: usize, ctx: &Ctx) -> Result<T, ()>
        where T: TryFromCtx<'a, Ctx> + ReadSize<'a, ctx>,
              Self: 'a;

}
```

### Implementation for `&[u8]`
```rust
impl<'a> Pread<'a> for &'a [u8] {
    fn pread<T, Ctx>(self: &'a [u8], offset: usize, ctx: &Ctx) -> Result<T, ()>
        where T: TryFromCtx<'a, Ctx> + ReadSize<'a, Ctx>,
              Self: 'a
    {
        let len = <T as ReadSize<'a, Ctx>>::read_size(&self[offset..], ctx);
        if self.len() < offset + len {
            Err(())
        } else {
            TryFromCtx::try_from_ctx(&self[offset..offset + len], ctx)
        }
    }
}
```

### Implementation for `&str`
```rust
impl<'a> ReadSize<'a, StrCtx> for &'a str {
    fn read_size(scroll: &'a [u8], ctx: &StrCtx) -> usize {
        match *ctx {
            StrCtx::Delimiter(delimiter) => scroll.iter().take_while(|c| **c != delimiter).count(),
            StrCtx::Length(len) => len,
        }
    }
}

impl<'a> TryFromCtx<'a, StrCtx> for &'a str {
    fn try_from_ctx(scroll: &'a [u8], _: &StrCtx) -> Result<Self, ()> {
        str::from_utf8(scroll).map_err(|_| ())
    }
}
```

**Comment by m4b** [comment]
I like how you rewrote my garbage multiline str delim function into this elegant single line: `StrCtx::Delimiter(delimiter) => scroll.iter().take_while(|c| **c != delimiter).count(),` :D

So, few things:

1. These are separate experimental implementations, correct?
2. I assume the error `()` is just for simple demonstration, and would be replaced with real error
3. I don't really like passing the Ctx by reference; to me, Ctx should be Copyable, simple struct, with minimal information. If it is complicated, then should pay the price of construction once, and in the `TryFromCtx` impl, use simpler ctx to fill the fields of the object one is constructing. Also, it penalizes developer by making them remember to type one more thing (the `&`), which is admittedly minor, but it adds up, and also penalizes reads/writes of integers by passing a pointer size instead of very likely a byte for the endian enum as is the most common case.

The first version is actually interesting, having TryFromCtx return a size; it sort of combines SizeWith with  with TryFromCtx (but at the cost of making implementation more complicated).  I believe this would still be derivable with a proc_macro, by gread local offset, then return *offset.

So Ive also been working on scroll experiments, and basically got a prototype completely generic Pread working at a basic level (must comment so many things though, everything is so interconnected and sensitive in scroll)!

```rust
pub trait MeasureWith<Ctx> {
    type Units;
    #[inline]
    fn measure_with(&self, ctx: &Ctx) -> Self::Units;
}

impl<Ctx> MeasureWith<Ctx> for [u8] {
    type Units = usize;
    #[inline]
    fn measure_with(&self, _ctx: &Ctx) -> Self::Units {
        self.len()
    }
}

impl<Ctx, T: AsRef<[u8]>> MeasureWith<Ctx> for T {
    type Units = usize;
    #[inline]
    fn measure_with(&self, _ctx: &Ctx) -> Self::Units {
        self.as_ref().len()
    }
}

pub trait Pread<Ctx, E = error::Error, I = usize>
 : Index<I> + Index<RangeFrom<I>> + MeasureWith<Ctx, Units = I>
 where
       Ctx: Copy + Default + Debug,
       I: Add + Copy + PartialOrd,
{
    #[inline]
    /// Reads a value from `self` at `offset` with the given `ctx`
    /// # Example
    /// ```rust
    /// use scroll::Pread;
    /// let bytes: [u8; 2] = [0xde, 0xad];
    /// let dead: u16 = bytes.pread_with(0, scroll::BE).unwrap();
    /// assert_eq!(dead, 0xdeadu16);
    fn pread_with<'a, N: TryFromCtx<'a, (I, Ctx), <Self as Index<RangeFrom<I>>>::Output, Error = E>>
(&'a self, offset: I, ctx: Ctx) -> result::Result<N, E> where <Self as Index<RangeFrom<I>>>::Output: 'a {
        let len = self.measure_with(&ctx);
        if offset >= len {
            // generic error here :/
            unimplemented!()
        }
        N::try_from_ctx(&self[offset..], (offset, ctx))
    }
```

So after being deleted in early prototypes, `MeasureWith` is back :laughing:

Couple lessons learned:

1. I am 99% sure I will delete Buffer module; it makes refactors just horrible, as the `Buffer` is used all over examples. I will do this first, so it's out of the way.  I will add the code to goblin for easier refactor/port to later scroll when that lands, and also as an api test in `api.rs`
2. Many of the tests in `lib.rs` are obsolete and make refactoring also awful.
3. I do believe the offset logic needs to be moved out of the `TryFromCtx` implementation, and just like `FromCtx`, the implementor can assume they can start reading from the beginning of the buffer, just not sure best way to do this that balances ergonomics, etc.

I really don't want to move away from generic Pread for `Self`; who knows how it will be used in the future. But perhaps it is too much, and should focus on `&[u8]` and make life easier; I'm not sure :/

**Comment by m4b** [comment]
Oh, forgot to add, I really like your StrCtx idea, and also agree that definitely need to get rid of `pread_slice` and `pread_unsafe`; these also contribute to really horrible refactor pains, but latter may only be possible with totally generic pread, or with `&[u8]` only version, not sure yet; have to balance many considerations here :/

Definitely `pread_slice` should probably go tho.  We will still be able to `pread_with::<[u8]>` though, yes? I ask because will make porting easier, and also sometimes it's nice to read out bytes given a len (say form binary) and have error instead of panic from indexing

**Comment by m4b** [comment]
So to summarize:

1. I will remove Buffer
2. Can start removal of `pread_slice` and add new StrCtx to accomodate
3. Optionally, remove some of the older Pread as a type parameter apis, because they make everything harder

After this, the API surface will be smaller, and experimental refactors will be easier.

**Comment by goandylok** [comment]
I'm not sure about how `MessureWith` works. Since implementor doesn't have access to slices, &str can't figure out where it is delimited.

I came to realize &Ctx is a bad idea. Just forget it XD

**Comment by m4b** [comment]
MeasureWith would be used on Pread Self. The str ctx would work because it would use the u8 API, and would read the length without offset just starting from beginning of bytes.  In principle I think could even make str ctx generic if felt like it, but probably not worth it hehe.

**Comment by goandylok** [comment]
I think that generic is not always a good stuff and may polute the code. We could generic slice type to `This` at least there are options other than &[u8] and we could generic index type to `I: AddAssign + Copy + Add + Default + Debug` at least someone uses index other than usize.

**Comment by goandylok** [comment]
So far, I can't experiment the refactor in `scroll` before removing the road blocks. So I set up the  repo [`scroll2`](https://github.com/goandylok/scroll2) for the experiment purpose.

Current stage: 
- `Pwrite` supported
- &str supported
- &[u8] supported
- no generic on index type (hard code as `usize`)
- no generic on slice type (hard code as `&[u8]`)
- error type is ()
- use the design of `TryFromCtx` in the first example -- returning size.
- no `MessureWith` invoved
- combine `Pread` with `Gread`, the same for `Pwrite` and `Gwrite`

---
## Quick View
I will give some snapshots of the code of the client side, which shows the usage difficulty.

### bytes
These are all the code needed to support read and write for bytes. Short enough and especially have cleaner api than original.
```rust
use prelude::*;

impl<'a> TryFromCtx<'a, usize> for &'a [u8] {
    #[inline]
    fn try_from_ctx(scroll: &'a [u8], len: usize) -> Result<(Self, usize), ()> {        
        if len > scroll.len() {
            return Err(());
        };

        Ok((&scroll[..len], len))
    }
}

impl<'a> TryIntoCtx<()> for &'a [u8] {
    fn try_into_ctx(self, scroll: &mut [u8], _ctx: ()) -> Result<usize, ()> {
        if self.len() > scroll.len() {
            return Err(());
        };

        scroll[..self.len()].clone_from_slice(self);

        Ok(self.len())
    }
}
```
### usage -- same as original
```rust 
let bytes = [0xde, 0xad, 0xbe, 0xef, 0x00, 0x00];
let read = bytes.pread(0, 4).unwrap();
assert_eq!(read, &[0xde, 0xad, 0xbe, 0xef]);

let mut bytes = [0; 4];
bytes.pwrite(0, read).unwarp();
assert_eq!(bytes, [0xde, 0xad, 0xbe, 0xef]);
```

**Comment by goandylok** [comment]
To summarize:
- The solution of returning size fits in the existing pread system and easy to implement.
- implementor have great responsibility to guarantee returning a correct size and bound check before read or write. Otherwise, all the following gread and gwrite will break.
- We should give up generic over underlying slice, the most important problem is one can not implement pread for any other slice outside this crate, execpt some types they defined by themselves. But we can generic over the item of slice, which allow scrolling around &[usize] or bits( &[bool]), but the impl constraint still exist; one need to newtype over primitives so as to impl TryFromCtx for them.
- Grouping pread and gread together help reduce api confusion.

**Comment by m4b** [comment]
@goandylok So I haven't had a chance to do a full review of the experimental form, but the approach is very exciting; I understand now what you mean by collapsing Gread into Pread via a single `TryFromCtx` whose size return is ignored for `pread` and `used` by `gread`; we can also add `Cread` into the mix, and have a method `cread` which uses a simple `FromCtx` or whatever it ends up being called.

I am worried at a high level that there are too many radical changes being planned, the new `TryFromCtx` being an example, and that is a little worrisome.

Another thing I noticed; won't the StrCtx length bounds check never fail? E.g., if the string (say null terminated) is longer than the buffer, it will just oob and panic, no?  Consequently, I don't think the check is even meangingful, unless I don't understand.  I think this is just another manifestation of variable length types versus statically known sized types, and i'm not sure the experimental version solves it.

Lastly, passing the sizeof the value in `pread` is just an example/typo right? the user should never have to pass the size of the integer.

In fact, the client API w.r.t. parsing only should not change at all, modulo warnings that `Gread` is unused/doesn't exist.

Client TryFromCtx will obviously break in this particular case.

**Comment by goandylok** [comment]
Yes, these are certainly radical break changes. That way I have to do it in new blank project. XD.

Actually, I didn't think about the case when str's delimiter never exist. But it is not hard to fix, we can simply return `Err(Error::InComplete)` when we reached the end.

The last example shows reading and writing byte slice from byte slice, which you mentioned before. So the number is a ctx, indicating the length of slice.

**Comment by m4b** [comment]
I think the error should be TooBig and we can pretty much get rid of all the others, iiuc. I realized now of course, the str either doesn't have delimiter and reaches end or the Length context is too big. In second case we can return TooBig since it's just a check against bytes len.

I think this might work... Wow so passing a number as ctx reads slices? Gosh that's cool! 

Ok. Interested to see this integrated, which will still be a big job. Best to start a branch soon. 

Do you think I should remove pread_slice first? It might even be easier to start integration by naming the trait `Pread2` and adding a `TryFromCtx` separate ctx, and then slowly replacing piecemeal. Also could benchmark implementations side by side. 

On that topic: name bikeshedding, some random thoughts:

1. ReadCtx
2. ReinterpretCtx
3. InfallibleReadCtx (FromCtx)
4. ReadWith
5. CtxRead

**Comment by goandylok** [comment]
I think the first thingb we need to do is to nail down API so as to clean up tests and benmarks.

So I am planing to copy old code to project scroll2 and iterate new code in scroll2 until we all satisfied with new design. Then, what we need to do is to copy-and-paste and write docs.

Ctx is good enough to me. :D

**Comment by goandylok** [comment]
The benchmark data interests me, where `cread` is 20 times faster than `pread`, which is really really weird because an unwrap and some extra bound check should not cost so much like this.
```
test bench_cread             ... bench:      53,826 ns/iter (+/- 10,911) = 3715 MB/s
test bench_pread_unwrap      ... bench:   1,294,288 ns/iter (+/- 983,546) = 154 MB/s
```
The more interesting is that it keeps this performance gap even after I made `TryFromCtx` only return Ok(0). That means the performance issues may not be caused by `TryFromCtx`. So I made a little change as follows.

```rust
let _: u16 = black_box(data.cread_with(0, LE));
// =>
let _: u16 = black_box(u16::from_ctx(&data, LE));
```
```rust
let _ = black_box(data.pread_with::<u16>(0, LE));
// =>
let _: u16 = black_box(u16::try_from_ctx(&data, (0, LE)).unwrap());
```
benchmark result
```
test bench_cread             ... bench:      52,599 ns/iter (+/- 5,179) = 3802 MB/s
test bench_pread_with_unwrap ... bench:      55,547 ns/iter (+/- 24,674) = 3600 MB/s
```
I can't tell why this happen right now, but I can see the opportunity to eliminate `FromCtx`. And I thought of some possibilities :
- `pread` involved dynamic dispatch
- `cread` have better strategy than `pread`

Anyway, this result definitely showed some performance problems of trait `pread`.

**Comment by m4b** [comment]
Are these coming from scroll itself or experimental repo ? Would need to see benchmark code; the current version needs better naming, there are actually two forms and one is always slower than the other. But it's actually a really good idea to also bench the Direct TryFromCtx impls. One thing that worries me is the new version now always returns a struct (tuple of value and size) which could be a bottleneck. It might be worthwhile to split the trait converters for your version with size returned and regular, needs examination. But don't optimize too crazy yet, design first :)

**Comment by goandylok** [comment]
This benmark comes from original scroll repo.

**Comment by goandylok** [comment]
It's not hard to benchmark on new design, defining a special function to wrap the try_from_ctx call in test module should be enough.

**Comment by goandylok** [comment]
Benchmark result from experiment repo
```
test bench_pread_with_unwarp  ... bench:      46,816 ns/iter (+/- 4,736) = 4272 MB/s
test bench_try_from_ctx       ... bench:      46,328 ns/iter (+/- 14,498) = 4317 MB/s
```

Benchmark code
```rust
#[bench]
fn bench_pread_with_unwarp(b: &mut test::Bencher) {
    const NITER: i32 = 100_000;
    b.iter(|| {
        for _ in 1..NITER {
            let data = black_box([1, 2]);
            let _: u16 = black_box(data.pread_with(0, LE).unwrap());
        }
    });
    b.bytes = 2 * NITER as u64;
}

#[bench]
fn bench_try_from_ctx(b: &mut test::Bencher) {
    const NITER: i32 = 100_000;
    b.iter(|| {
        for _ in 1..NITER {
            let data: &[u8] = &black_box([1, 2]);
            let _: u16 = black_box(u16::try_from_ctx(&data, LE).unwrap().0);
        }
    });
    b.bytes = 2 * NITER as u64;
}
```

**Comment by m4b** [comment]
So on my machine modifying above for current scroll, i basically see no difference:

```
test bench_pread_with_unwrap_2 ... bench:      40,207 ns/iter (+/- 1,399) = 4974 MB/s
test bench_try_from_ctx_2      ... bench:      40,163 ns/iter (+/- 573) = 4979 MB/s
```

```rust
#[bench]
fn bench_pread_with_unwrap_2(b: &mut test::Bencher) {
    use scroll::{LE, Pread};
    const NITER: i32 = 100_000;
    b.iter(|| {
        for _ in 1..NITER {
            let data = black_box([1, 2]);
            let _: u16 = black_box(data.pread_with(0, LE).unwrap());
        }
    });
    b.bytes = 2 * NITER as u64;
}

#[bench]
fn bench_try_from_ctx_2(b: &mut test::Bencher) {
    use scroll::{LE, Pread};
    use scroll::ctx::TryFromCtx;
    const NITER: i32 = 100_000;
    b.iter(|| {
        for _ in 1..NITER {
            let data: &[u8] = &black_box([1, 2]);
            let _: u16 = black_box(u16::try_from_ctx(&data, (0, LE)).unwrap());
        }
    });
    b.bytes = 2 * NITER as u64;
}
```

I don't know if it's worthwhile to start getting super into benches right now without comparing both things side by side on same machine.

**Comment by m4b** [comment]
Cread reports basically the same for me:

```
test bench_cread_with_2        ... bench:      40,248 ns/iter (+/- 1,261) = 4969 MB/s
test bench_from_ctx_2          ... bench:      40,155 ns/iter (+/- 254) = 4980 MB/s
```

```rust
test bench_cread_with_2        ... bench:      40,245 ns/iter (+/- 4,257) = 4969 MB/s
test bench_from_ctx_2          ... bench:      40,206 ns/iter (+/- 592) = 4974 MB/s
```

**Comment by m4b** [comment]
That's strange your unwrap version was so different; does it consistently underperform at 20x on your machine?

**Comment by goandylok** [comment]
I need to dig deep into benchmark now because this help make decision that whether I keep the trait `FromCtx`, since it exists just to fix performance issue.

I re-pulled the whole repo and did benchmark again. The result still shows significant differences.
```
test bench_byteorder           ... bench:      46,801 ns/iter (+/- 3,962) = 4273 MB/s
test bench_byteorder_vec       ... bench:   6,613,969 ns/iter (+/- 1,341,831) = 151 MB/s
test bench_cread               ... bench:      47,036 ns/iter (+/- 3,798) = 4252 MB/s      <- Here
test bench_cread_vec           ... bench:   1,167,389 ns/iter (+/- 97,044) = 856 MB/s
test bench_gread_unsafe        ... bench:      46,728 ns/iter (+/- 6,172) = 4280 MB/s
test bench_gread_vec           ... bench:   5,839,875 ns/iter (+/- 272,070) = 171 MB/s
test bench_parallel_cread_with ... bench:     508,134 ns/iter (+/- 239,940) = 1967 MB/s
test bench_parallel_pread_with ... bench:   2,546,503 ns/iter (+/- 2,005,598) = 392 MB/s
test bench_pread_ctx_vec       ... bench:   6,474,964 ns/iter (+/- 485,475) = 154 MB/s
test bench_pread_unsafe        ... bench:      47,163 ns/iter (+/- 3,142) = 4240 MB/s
 test bench_pread_unwrap        ... bench:   1,080,362 ns/iter (+/- 93,143) = 185 MB/s      <- And here
test bench_pread_vec           ... bench:   6,634,524 ns/iter (+/- 191,867) = 150 MB/s
test bench_pread_with_unwrap   ... bench:      47,776 ns/iter (+/- 4,860) = 4186 MB/s
```

**Comment by m4b** [comment]
Yea that's significant; what CPU/OS are you running, just curious?

Remember that something like `FromCtx` will have to exist to allow `lread`, as `Read` trait forces essentially copy or owned values, unless other tricks are played.

I'm going to try the generic patch again; i think it has to be done all at once, and slice can't be fixed until it's added, for various reasons

**Comment by goandylok** [comment]
I am using Surface Pro 4 with Window10.
CPU Intel(R) Core(TM) m3-6Y30 CPU @ 0.90GHz 1.15 GHz
RAM 4.00 GB

**Comment by goandylok** [comment]
I don't think `Read` a roadblock of removing `FromCtx`. In the case of requiring a concrete and owned value, we can add a 'static bound like `T: TryFromCtx<_, _> + 'static`.

**Comment by m4b** [comment]
> I don't think Read a roadblock of removing FromCtx

If you think that will work, very interested!

The benchmark issue is interesting... I'll paste mine here for comparison later:

```
test bench_byteorder           ... bench:      40,205 ns/iter (+/- 2,819) = 4974 MB/s
test bench_byteorder_vec       ... bench:   1,431,136 ns/iter (+/- 31,749) = 698 MB/s
test bench_cread               ... bench:      40,225 ns/iter (+/- 1,050) = 4972 MB/s
test bench_cread_vec           ... bench:   1,005,121 ns/iter (+/- 5,124) = 994 MB/s
test bench_cread_with_2        ... bench:      40,248 ns/iter (+/- 1,261) = 4969 MB/s
test bench_from_ctx_2          ... bench:      40,155 ns/iter (+/- 254) = 4980 MB/s
test bench_gread_unsafe        ... bench:      40,180 ns/iter (+/- 352) = 4977 MB/s
test bench_gread_vec           ... bench:   1,205,387 ns/iter (+/- 22,093) = 829 MB/s
test bench_parallel_cread_with ... bench:     323,235 ns/iter (+/- 61,368) = 3093 MB/s
test bench_parallel_pread_with ... bench:     793,697 ns/iter (+/- 45,438) = 1259 MB/s
test bench_pread_ctx_vec       ... bench:   1,214,684 ns/iter (+/- 12,617) = 823 MB/s
test bench_pread_unsafe        ... bench:      40,201 ns/iter (+/- 454) = 4975 MB/s
test bench_pread_unwrap        ... bench:     951,712 ns/iter (+/- 19,128) = 210 MB/s
test bench_pread_vec           ... bench:   1,207,566 ns/iter (+/- 17,949) = 828 MB/s
test bench_pread_with_unwrap   ... bench:      40,145 ns/iter (+/- 777) = 4981 MB/s
test bench_pread_with_unwrap_2 ... bench:      40,210 ns/iter (+/- 434) = 4973 MB/s
test bench_try_from_ctx_2      ... bench:      40,183 ns/iter (+/- 477) = 4977 MB/s
```

**Comment by m4b** [comment]
Weird, this just randomly repro'd for me when refactoring:

```
test bench_byteorder           ... bench:      40,163 ns/iter (+/- 362) = 4979 MB/s
test bench_byteorder_vec       ... bench:   1,430,539 ns/iter (+/- 20,509) = 699 MB/s
test bench_parallel_pread_with ... bench:     787,579 ns/iter (+/- 49,706) = 1269 MB/s
test bench_pread_ctx_vec       ... bench:   1,207,051 ns/iter (+/- 18,367) = 828 MB/s
test bench_pread_unwrap        ... bench:     952,080 ns/iter (+/- 7,643) = 210 MB/s
test bench_pread_vec           ... bench:   1,207,542 ns/iter (+/- 18,211) = 828 MB/s
test bench_pread_with_unwrap   ... bench:      40,180 ns/iter (+/- 409) = 4977 MB/s
```

note the `with_unwrap` is not slow at all

**Comment by m4b** [comment]
Here is why i want a generic interface to pread, you simply write this:

```
impl<Ctx: Copy + Default + Debug,
     E,
     I: Add + Copy + PartialOrd,
     R: ?Sized + Index<I> + Index<RangeFrom<I>> + MeasureWith<Ctx, Units = I>>
    Pread<Ctx, E, I> for R {}
```

you don't repeat yourself by rewriting the method signatures and the corresponding impl, you don't write anything more, and if/when you refactor, you hardly ever change that part.

Unfortunately, much to my chagrin, I just realized that moving the offset logic into a generic `Pread` might be impossible; consider:

```rust
    fn pread_with<'a, N: TryFromCtx<'a, Ctx, <Self as Index<RangeFrom<I>>>::Output, Error = E>>(&'a self, offset: I, ctx: Ctx) -> result::Result<N, E> where <Self as Index<RangeFrom<I>>>::Output: 'a {
        let len = self.measure_with(&ctx);
        if offset >= len {
            // generic error here / what do I return?
            panic!("offset >= len")
        }
        N::try_from_ctx(&self[offset..], ctx)
    }
```

it seems the only solution to this problem is to force a single error to be used (Scroll's error), but I'm not sure this is entirely reasonable, or a good feature.

@goandylok do you think there is a way to solve this problem?

**Comment by m4b** [comment]
The more I think about this, the more this is a huge problem, and might prevent moving offsetting logic into `pread`... :/ I just realized in all your examples, you hardcode the error in the type signature; I guess we could force all clients to use a single error emanating from scroll, but that kind of sucks :/

I've pushed the WIP branch, `generic_pread`; i'd encourage you to start testing some of your proposals on there and see how it interacts with all the complicated pieces.

So `gread` module is disabled, `leb128.rs` module disabled, along with `api` tests, `gread` tests in `lib.rs`.

The `TryFromCtx` traits are updated to remove tuple logic, along with generic `Error` type as well.  This should set the stage for easier integration and testing.

The tests can even be run, but `str` fails do to offset assertion firing because of above problem

**Comment by goandylok** [comment]
My opinion is not to repeat generic signature but to remove generics over index type and slice type. 

My next step on experiment is error type. My initial idea is to provide a Error type that have generic argument, which may solve your problem.

```rust 
enum Error<E: Error> {
    OffsetEcceed,
    TooBig,
    Custom(E)
}
```

**Comment by goandylok** [comment]
I think I will start porting codes to that branch and see how they interacte with existing code base after I add error support and lread.

**Comment by goandylok** [comment]
> you don't repeat yourself by rewriting the method signatures and the corresponding impl, you don't write anything more, and if/when you refactor, you hardly ever change that part.

```rust
pub trait Pread {
    fn pread_with<'a, T, Ctx>(&'a self, offset: usize, ctx: Ctx) -> Result<T, ()>
        where T: TryFromCtx<'a, Ctx>;
    // ...
}
```

I have just understood what you means here :-o , this is a design mistake to repeat the signature again and again. It's because I filled the traits one function by one function, thus I didn't notice that they had the same generic types. But it is not about refactor, what you said it's is hard to change the code, it about api interface.

In the current version, when `pread`, user is forced to provide type T as long as its Ctx regardless whether type Ctx is known by compiler.
```rust 
assert!(bytes
             .pread_with::<&str, _>(0, StrCtx::Length(27))
             .is_err());
```

I made a small patch to change the `pread` trait as following, which will help improve ergonomic.

```rust
pub trait Pread<'a, Ctx> {
    fn pread_with<T>(&'a self, offset: usize, ctx: Ctx) -> Result<T, ()>
        where T: TryFromCtx<'a, Ctx>;
    // ...
}
```
Note that I retained type `T` in function signature, this is because if we don't, user will have to trait-cast the `Pread` to provide type annotation.
```rust 
assert!(bytes
             .pread_with::<&str>(0, StrCtx::Length(27))
             .is_err());

// versus

assert!(Pread<_, &str, _>::pread_with(&bytes, 0, StrCtx::Length(27)).is_err());
```

**Comment by m4b** [comment]
Haven't had a chance to read these comments yet but was given a suggestion over on internal forums for the generic error problem; pretty obvious on hindsight - use more generics!

The error type now has a `E: From<error::Error<I>>` bound, which allows one to call `into()`; very elegant solution. I believe full generic version can work now.

**Comment by goandylok** [comment]
> The error type now has a `E: From<error::Error<I>>` bound, which allows one to call into(); very elegant solution. I believe full generic version can work now.

This is not idiom for rust, as `std::io::Error` and others error type in std don't have `From<>` implementation. Instead, rust provides auto error-casting by question mark by calling `.into()` in code expand. So I vote against this design.

**Comment by goandylok** [comment]
I think we have disagreement about slice generic, when I think it brings too many troubles and you think this make scroll work on wider domain. But the problem is whether we have options, other than byte slices? Can user define one in their systems? Are there any use case that is worth paying the price, paid by both author and user?

**Comment by goandylok** [comment]
> I don't see the point of adding another generic type parameter to Error; it would only be used by us, 

Yahh.. You are right. We should not place generic types on Error because it blocks user from storing and popping error of different type.

Tho, The semantic of `len` parameter of `Error::TooBig` is not clearly defined thus useless and it will be a tricky part when trying to port returning size in. So I suggest removing it.

I am going to work on returning size but I've found some issues need to discuss before that. 
- `Iter` is now available and seems better than array solution. Whether you want to adopt it?
- We need to count delimiter into size but exclude from content if we support `Iter`.
- ByteCtx has landed. Whether to adopt it?
- Remove Generic From bound of `Error` due to auto-cast mechanism of question mark.

## Iter
```rust
#[test]
fn test_iter() {
    let bytes: &[u8] = b"hello\0world\0dead\0beef\0more";
    let mut offset = 0;
    {
        let mut iter = bytes.gread_iter(&mut offset, StrCtx::Delimiter(NULL));
        assert_eq!(iter.next(), Some("hello"));
        assert_eq!(iter.next(), Some("world"));
        assert_eq!(iter.next(), Some("dead"));
        assert_eq!(iter.next(), Some("beef"));
        assert_eq!(iter.next(), None);
    }
    assert_eq!(offset, 22);
}
```

## ByteCtx
```rust
#[test]
fn test_bytes_pattern() {
    let bytes: &[u8] = b"abcde\0fghijk";

    assert_eq!(TryRead::try_read(&bytes, ByteCtx::Length(4)).unwrap(),
               (&b"abcd"[..], 4));

    assert_eq!(TryRead::try_read(bytes, ByteCtx::Pattern(b"abc")).unwrap(),
               (&b"abc"[..], 3));

    assert_eq!(TryRead::try_read(bytes, ByteCtx::UntilPattern(b"fg")).unwrap(),
               (&b"abcde\0fg"[..], 8));

    assert_eq!(TryRead::try_read(bytes, ByteCtx::UntilPattern(b"jk")).unwrap(),
               (&b"abcde\0fghijk"[..], 12));
}
```

**Comment by goandylok** [comment]
## Proposal: Unifying pread and gread

### API
```rust
// work at the same time
bytes.pwrite_with(0, num, LE).unwrap();
bytes.pwrite_with(&mut 0, num, LE).unwrap();
```

### Signature
```rust
fn pwrite_with<I, T>(&mut self, offset: I, t: T, ctx: Ctx) -> Result<()>
        where T: TryWrite<Ctx>,
              I: Index
```

### Trait
```rust
pub trait Index {
    fn get(&self) -> usize;

    fn set(&mut self, value: usize);
}

impl<'a> Index for &'a mut usize {
    fn get(&self) -> usize {
        **self
    }

    fn set(&mut self, value: usize) {
        **self = value;
    }
}

impl Index for usize {
    fn get(&self) -> usize {
        *self
    }

    fn set(&mut self, value: usize) {}
}
```

## Drawback
One may forget to write &mut before offset, which may result in bugs -- not increasing the offset.

**Comment by goandylok** [comment]
How do you work with the current version scroll, without returning size, to know what the next offset is?
You can `pread` only one `str` from slice then no more following stuffs because you don't know where the next starts.

**Comment by m4b** [comment]
> Tho, The semantic of len parameter of Error::TooBig is not clearly defined thus useless and it will be a tricky part when trying to port returning size in. So I suggest removing it.

Yea, I agree now, it's not clear what it means exactly; I do think we should still return the size of the thing we wanted to return. I'm just hesitant to make error messages less informational; they were extremely helpful to me when the offset was present...

> How do you work with the current version scroll, without returning size, to know what the next offset is?
You can pread only one str from slice then no more following stuffs because you don't know where the next starts.

Yes, this is true. `&str` doesn't impl `SizeWith` so it can't use `Gread`; this is intentional.  So it's up to the user to appropriately read the next str, like `pread::<&str>(offset+last.len() +1)` or something

> Iter is now available and seems better than array solution. Whether you want to adopt it?

I'd like to wait on this. But --- this is a _really_ cool idea, and I'd like to experiment with this more, but since it will be a method, I believe we can add it later. If we made this generic, I could do something like `bytes.pread_iter::<ProgramHeader>(offset)?.collect::<Vec<ProgramHeader>>()` which is so cool :heart_eyes_cat: 

> We need to count delimiter into size but exclude from content if we support Iter.

I think most people would expect this behavior, no? On a similar note, I am constantly bitten by this when reading _and_ writing null terminated strings, as I routinely forget the +1 delimiter, so I think this would be welcome.  I can't think of any usecases where you wouldn't want to count the delimiter when incrementing a size to read at _next_, no?

> ByteCtx has landed. Whether to adopt it?

I like this; I'm wondering if we can make it more generic though? E.g., something like `SliceCtx` which reads out arbitrary slices of objects?

> Remove Generic From bound of Error due to auto-cast mechanism of question mark.

I'm not sure we can do this, but we can try.

---------------------------------------

Your proposal for pread/gread union is very very interesting --- again! :) :+1: 

I'm definitely intrigued.  For my uses, I usually do this in `TryFrom/Into` definitions:

```
                    let mut offset = &mut offset;
```

But probably I only wrote this because the compiler wouldn't let me call `gread` with the regular offset, in which case I would have forgotten and puzzled about for a while wondering what went wrong.

On the other hand, it is really elegant and convenient; I think we can punt on this idea for a bit as well.

For now, I think it's really important to get the following hammered down, as it lays the foundation for really great, new and unified API:

1. getting the size returning variants working for `TryFromCtx` and `TryIntoCtx`
2. fold `gread, lread, cread` into single `Pread` trait, which should be possible now due to generics and 1. (this is a _HUGE_ usability win in my opinion; I'm one of the few where less imports is better imho :))
3. Benchmark (I don't think there will be any major difference, but returning a struct/tuple with size is pretty different)
4. Lastly, bikeshed on new names for `TryFromCtx` and `TryIntoCtx`; (I'm currently partial to `TryReadWith` and `TryWriteWith`, and `ReadWith`/`WriteWith` for `FromCtx`/`IntoCtx`)
5. `FromCtx` should now allow auto-deriving, which will be very handy

So, I think it's time to open a new PR trying out 1. and 2. above!

**Comment by goandylok** [comment]
> I like this; I'm wondering if we can make it more generic though? E.g., something like SliceCtx which reads out arbitrary slices of objects?

This is what the `Iter` for. Slice has many downsides. The first is lifetime; it's impossible to create one inside a function; it must be passed in by ctx. The second one is capacity; how do we guarantee the length of slice is larger than we need? But `Iter` easily bypassed these problem and provided more functionalities and flexibility. For example, `bytes.gread_iter().skip(1).take(5).collect()`

**Comment by m4b** [comment]
I've been refactoring goblin for over 3 hours.

So:

1. not doing this again :)
2. scroll has definitely lost functionality -
   a. pread_slice is missing (this is more helpful than i realized, it prevents bound check errors which would be panic; instead allows to be properly propagated as error); also for strs it becomes very verbose and requires another import which is annoying (when refactoring)
   b. can no longer pread out of `AsRef<[u8]>` generics; this is pretty sad :/ - i don't know if i would have even began the generic refactor if i knew this would happen
3. I'm not even sure why I need to refactor; don't seem to have gained anything, except missing functions. The `TryOffsetCtx` without offset is nicer when impling, but it's just burdensome to refactor, I think this was a good thing.
4. I don't see any of the benefits yet; e.g., a single trait Pread to import with all the methods.

Overall I think the changes are good, but I think need to stop breaking API and focus on stablizing and getting the other features in asap

**Comment by m4b** [comment]
E.g., this once very pleasant and nice code without any error checking:

```
        if let Some(member) = self.get(member) {
                let bytes = buffer.pread_slice(member.offset as usize, member.size())?;
                Ok(bytes)
            }
```

 becomes extremely tedious and very easy to mess up:

```
        if let Some(member) = self.get(member) {
            let len = member.size();
            if member.offset as usize + len >= buffer.len() {
                Err(Error::Malformed(format!("Cannot extract member {:?} -- too big", member).into()))
            } else {
                let bytes = &buffer[member.offset..member.offset+len];
                Ok(bytes)
            }
```

**Comment by goandylok** [comment]
That's because you start refactor on downstream crates before returning size being landed. XD. All this features are design around returning size and there are new stuffs based on these changes. 

## Talk is cheap
Assume we defined a custom type `Header`, which have a varibal-length name and a `bool` field. 

### Byte Representation
```
|       | Length of name (Big Endian) |                Name              | Enabled |
| ----- | --------------------------- | ---- | ---- | ---- | ---- | ---- | ------- |
| Byte  | 0            | 5            | 'H'  | 'E'  | 'L'  | 'L'  | 'O'  | 0       |
```

```rust
use byte::*;
use byte::ctx::*;

struct Header<'a> {
    name: &'a str,
    enabled: bool,
}

impl<'a> TryRead<'a, Endian> for Header<'a> {
    fn try_read(bytes: &'a [u8], endian: Endian) -> Result<(Self, usize)> {
        let offset = &mut 0;

        let name_len = bytes.read_with::<u16>(offset, endian)? as usize;
        let header = Header {
            name: bytes.read_with::<&str>(offset, Str::Length(name_len))?,
            enabled: bytes.read(offset)?,
        };

        Ok((header, *offset))
    }
}

impl<'a> TryWrite<Endian> for Header<'a> {
    fn try_write(self, mut bytes: &mut [u8], endian: Endian) -> Result<usize> {
        let offset = &mut 0;

        bytes.write_with(offset, self.name.len() as u16, endian)?;
        bytes.write(offset, self.name)?;
        bytes.write(offset, self.enabled)?;

        Ok(*offset)
    }
}
```

### Usage

```rust
let bytes = [0, 5, b"H"[0], b"E"[0], b"L"[0], b"L"[0], b"O"[0], 0];

// read
let header: Header = bytes.read_with(&mut 0, BE).unwrap();
assert_eq!(header.name, "HELLO");
assert_eq!(header.enabled, false);

//write
let mut write = [0u8; 8];
write.write_with(&mut 0, header, BE).unwrap();
assert_eq!(write, bytes);
```

**Comment by m4b** [comment]
> That's because you start refactoring on downstream crates before returning size being landed. XD

Must. Refactor. Crates. :D

Yea your example is perfect usecase mixing variable with fixed size and auto incrementing along the way

I think this means can also drop SizeWith trait from proc macros and impls which is cool. It always felt like stupid boilerplate.

Feel free to open up a PR; I just have this bad feeling well get almost there and then something will come up and make it infeasible :laughing: (this has happened to me a couple of times while working on scroll, hence I got paranoid about API tests; and it shows cause AsRef magically unimplemented itself.)

**Comment by goandylok** [comment]
In fact, I wish to publish a extremely Stripped-down version of scroll, for the use in embedded hardware. But there will be duplication of functionalities. What's your opinion on this? Working two projects in parallel and sharing codes?

**Comment by goandylok** [comment]
I think the final code will be push to the [experiment repo](https://github.com/goandylok/byte) these days.

**Comment by goandylok** [comment]
After that I think we should discuss whether to publish that and open pulls on scroll or something else.

**Comment by m4b** [comment]
I was afraid something like this would happen :(

You are of course free to do whatever you like.

Here are some reasons not to:

1. It's basically a hard fork without any of the reasons for hard forking - developer not responding, features not being adopted, etc.
2. Because it's basically a hard fork, almost all the functionality is duplicated, and all of the code is repeated
3. I don't see any reason why the `no_std` version of scroll doesn't do exactly what you want
4. Of course it's easier to work by yourself - initially - but remember the state of the crate you're using is built on top of all the work that got scroll to that point. Consequently, I'd urge you to add to that base, instead of creating an entirely new history and splitting developer time.
3. I literally do not have time to maintain and look at code in two crates with basically duplicated functionality
4. We've spent a great deal of time and energy collaborating and already adopting many of your ideas, and I think I've been more than responsive, but now you are thinking of just publishing your own crate (why didn't you say this before, I am very busy, and I wouldn't have started _any_ of these refactors - this is truly the most vexing part of all)
5. As far as I can see I don't see any serious difference between your crate and scroll except you changed names and added the size returning functionality (which is a great idea) that started this whole thing. Why didn't you just start a PR doing this like I asked about a week ago, and it would essentially be done?
6. It fragments any incoming issues or developers between the two crates; why, what's the purpose, again, they're basically duplicating functionality?
7. You have collaborator rights for this repo, so again, it's not like I haven't been responsive, or you haven't been able to experiment in branches on this repo.

I could go on; I think most developers would be pretty pissed at this point to be honest, especially after all this work.

I don't mean to be harsh, and again it has been very pleasurable working with you, but just letting you know I think basically forking, copying and publishing a slightly different scroll with a different name at this point is not a very nice thing to do, and I would strongly urge you to continue collaborating instead of doing this, as I think we can work through any issues that arise.

A new crate duplicating functionality, and a hard fork is not the answer.  That is my position, and I truly hope you'll reconsider :)

**Comment by goandylok** [comment]
I wish I had explained details, so it wouldn't lead to misunderstand. 

I republished the second version scroll is not for any selfish reasons. But because some significant differences that I considered impossible to merge in current scroll. For example, `FromCtx` was abandoned since lread was not required by embedded systems. And pread along with pfile was gone for the same reason. I think it is, in your words, too radical for scroll. But I am trying to port the new, acceptable functionality to scroll.

In the other aspects, the experiment repo is not based on scroll. Insted, it's based on a demo project I design before, as I mentioned sometime, scroll has exactly same api with my draft. And I also learned some awesome ideas from scroll, such as increasing offset by usize references.

In the long progress, I finally realized scroll has different design target -- scroll wants to be generic when I wants to be specific and small. This leads to the new crate.

**Comment by goandylok** [comment]
The close button always disturbes me.

I am not object to yank the new crate, and port codes into scroll -- that's what I planned to do as I said before, design in experiment repo then copy-and-paste. But do scroll accept these radical refactoring? Move Leb128, lread into other crate; unify gread and pread by removing pread(laugh) (it's because pread became useless after returning size); remove cread.

Again, I am not intend to make my own crate, but I am not clear these refactoring could happen in scoll, as your concerns on generic or error or lread, etc. It's not too late to reconsider this choice  , for me.

**Comment by m4b** [comment]
I would still strongly urge you to consider collaborating as two instead of two separate projects. There is strength in numbers. 

And again I still don't see why you need another crate; this just doesn't follow for me. 

Lread is under std feature flag and not even present in no_std environment. 

Similarly I suspect you'll eventually want a cread like nonfailing reader writer at some point anyway, and even if not, those methods won't be in you're binary if you don't use them, so what does it matter if they're provided ? Are you concerned about binary bloat? Size of source code? Compile time? None of these are large or serious issues. 

What if you want to read bytes out of your embedded device at a different endianness and write them to a file on your host machine ? You'd have to use scroll which has the same API basically; but why, just have all the same functionality in one crate and then it's all tested and a cohesive whole instead of fragmented. 

As for genericism, I don't see how it affects your design goal - you just want to read bytes out of a buffer, what does it matter that the trait is generic ?

So I just don't understand the need for a crate which duplicates 2/3 of scroll, but which hand reifies the byte type.

Perhaps im missing something but im definitely willing to work closely with you to get scroll to a state where it's exactly what you need for embedded and exactly what anyone needs for just reading and writing bytes easily.

I just think having another crate is a mistake and fractures potential adopters and hence bug reports and new design considerations needlessly. 

Honestly you could even just wrap scroll in your crate, force no std, export only your methods, inline always everywhere and it would accomplish the same thing I think, with 90% less code and a single source of truth/tested code bodies/bugfixes

**Comment by m4b** [comment]
It's hard for me to evaluate anything without a PR. As we planned I'd like to see the offset returning logic added to scroll

Adding or removing methods can be evaluated after that. Many of your concerns about leb128 etc are easily solved by feature flags

**Comment by m4b** [comment]
And I don't think yanking a crate is necessary. As I said if you want a stripped version of scroll you could just have your methods and use scroll behind scenes, etc. 

The reason I'm against two crates (whether that's scroll or your crate) is it fractures issues and users for very little gain. I'd prefer to have one slightly more complicated crate and have all the users and bug reports and two developers addressing it than split between the two, that's all :)

**Comment by goandylok** [comment]
The problem is learning curve. Byte encoding and decoding is an important missing piece of rust ecosystem on embedded system. But the complexity of scroll, including lread, generic signature, mixed namespaces
 and lack of runtime offset would keep users from willing to learn it. So I propose d to improve this crate, like renaming, module restructuring, specializing for slice, returning size, removing some scattered things.

So what you suggested that let Byte depend on scroll is somehow feasible. But I still wish they could be in one. I really don't care who or how to own the crate, but I want there is one, which is easy to use, easy to learn.

**Comment by m4b** [comment]
Well it sounds like we're in agreement on basics about single crate. Some of the other issues I agree and disagree with but that's fine :) especially agree about learning curve.

I do think the genericism may hurt but first thing people look at are examples, which there are plenty. Here I think scroll is easily the easiest to use crate there is. I don't see it getting easier. And then probably TryFromCtx impl. Latter has been made better without offsets. We will change name, I agreed to this already.

IO reading is very important and I don't think this should be understated. I regret actually adding it so late. I propose to rename this ioread. Once all methods are in a single Trait I think this will be very clear and easy for newcomers to see what to use. 

Many other things are negotiable

**Comment by goandylok** [comment]
Forgot to say, there is an important reason for the new crate, it's so important that leads me to chose to work in two project in parallel, the panic. Hardware can't endure any possible panic, so that if we want to have only one crate, we have to hide all codes with assert behind features flag, especially the cread.

**Comment by m4b** [comment]
You explicitly index byte slices. This can panic.

**Comment by goandylok** [comment]
If byte slices is indexed before check, it's a bug. If cread panics when things too big, that's what it is designed to be.

**Comment by m4b** [comment]
I still don't see the problem. So it panics if you use a slice too small. Don't use it unless you absolutely know its big enough. Not asserting Otherwise it would be UB reading. Why is this a problem for embedded ? I'm sure if I disassemble your code for an embedded target it will still contain indexing code which can technically abort/panic. 

But again, even if this is an issue just having it around, a feature flag would solve all those problems.

**Comment by m4b** [comment]
And again, if you don't use it any sufficiently smart linker will not contain that code, so I don't see how it's even a problem exposing it. Some embedded devs I'm sure will want a resultless api that they use for arrays whose lengths they absolutely know statically cannot panic.

**Comment by m4b** [comment]
There are lots of solutions to this problem, is my point. Off the top of my head:

1. Not care/ document
2. Blanket feature flag
3. Feature flag "unsafe" which adds unsafe to Cread and removes assert check and adds documentation explaining how it's dangerous but has no bounds checks for the speeds

Etc. I actually like 3.

**Comment by goandylok** [comment]
> Feature flag "unsafe" which adds unsafe to Cread and removes assert check and adds documentation explaining how it's dangerous but has no bounds checks for the speeds

I think this is not a good idea. While unsafe block marks a block unsafe, a unsafe flag marks the whole library unsafe.

I have reconsidered last night, and then I will talk about my opinion.

1. `byte` have different goals from `scroll`. `byte` needs to be safe, to be panic-free, from a user's point of view. While `scroll` wants IO and operation system. Indeed, we can have these goads in one, but it brings feature flags to gate each other. In fact, if you look into the source code of two crate now, you will find that they don't share too many code except `TryFromCtx` definition and Error type. But it will become more after we port returning size into scroll.

2. `scroll` is too complex. This is about learning curve and maintainability. If you think `scroll` can go no more easier, it's `byte`, that's what `scroll` can be. In the process of refactoring, we all felt the pressure from scattered code in the corner. There are no way to change them one by one, as you would start refactoring `goblin` after `scroll` is stabilized. At last, we had to comment out bunch of code to clear the road. And now, the more feature flags, the more complex it goes.

3. `byte` is actually the core part of final `scroll`, so here comes a crazy idea: make `scroll` depend on `byte` since `byte` is less than `scroll`, then provide leb128, system io, cread, file operations in `scroll`. It could be better than mixing all things into one, and stripped parts by feature flag every time. This requires us to have consensus on core design (mainly the generic part). On this point, I can even transfer the ownership of `byte` to you -- it is easier to be maintained by same person.

**Comment by m4b** [comment]
I think 3. might be a potential solution, but I don't know how it will work. It sounds like more refactors (done by who?)...

I don't want ownership of byte; I just don't have time to maintain another crate like this.

As for core ideas, I don't think abandoning generics is smart at all; at least not for looking to future.  Once specialization + const lands, I believe will be able to add const generic impls going all the way down to `TryFromCtx` also taking `[u8; 10]`, etc.

As for 2., I just don't agree with this at all on both points. When it comes to learning, _using_ scroll is about as easy as it gets.  It's easier than byteorder by a long shot imho.  With a single crate, I can write data into files, byte slices, vectors, arrays, all generic over the type, which makes refactors easy. And with proc macros, I can make it automatic.  I can guarantee most people who use byte order the first time do not understand at all why they have to write `bytes.read_u32::<BigEndian>()` and also import 3 separate items from the crate.  But it doesn't matter; they see the examples, they copy paste them, and then the understand it later (maybe).

Similarly, here is the type signature of partition: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition

Most beginners have no idea wth that means or why it is so long.  But they don't need to; how easy it is to _use_ is all that matters.

As for maintainability, again, I don't see this.  Removing the offsets from `TryFromCtx` was an _enormous_ api change, both internally _AND_ externally; it breaks all downstream clients (_this_ was why the goblin refactor was bad, not missing `pread_slice`).  If I knew anyone was using scroll, I probably wouldn't do it; it's really actually bad behavior to break something at that level without a major api change (even in 0.* versions).  If after adding feature after feature for several months on end to `byte`, and then you suddenly change the core functionality, I think that you will see that you end up commenting out tests, modules, and everything else in order to get it working again - this is just what a major refactor looks like...

The more you develop on a crate the more features you have which get added for a reason. `byte` is a new crate, basically rewrite from scratch, with many ideas and structures taken wholesale, which were refined elsewhere.  Many of the numeric transformation logic was directly inspired by byteorder, and I'm sure they spent a large deal of time refining that. But starting from scratch with templates already built is the easy part.  This is a well known phenomenon in software engineering. Once you have a structure in place and start making changes, fixing bugs, then you will feel the same pain, that's just how things are, and that's just how incremental development is, most things get added for a reason; but you're also building on top of something and making it better.

For example, it seemed ok to remove `pread_slice`; but after refactoring, I suddenly realized why I added it in the first place, and it was a good thing - less code, less manual offsets, less error checking.

And for the record, I started refactoring goblin to use the new scroll api, because I wanted to see how the changes looked in practice and how a client would actually use them; this is often much more important than how the library looks theoretically.  Many of the functions that got added to gread seem redundant, but they're extremely useful when doing all sorts of parsing, which makes the programmers life easier, which is all scroll is for.  I don't care if the implementation is complicated; i care if it's easy to use (and also fast in this case).

Yes scroll is complicated; but I don't think it's bad.  It may be harder to maintain, but again, I think any sufficiently complicated crate that's been battle tested will start to have warts.  `goblin` is a _huge_ pain to refactor; I've done it many times.  But that's also because it's battle tested and the complication is there for reason.

As for 1., I've already addressed these points, and I don't understand them, especially panic free notion.  You can panic on overflowing integers in rust.  There are hundreds of operations in the core lib that can panic. That `byte` is fundamentally different because scroll exposes something like `cread` just doesn't make sense to me.  It's a library.  You give clients as many options as possible.  Some may want to use panic version. Some may want to unwrap result. If they don't use it; it won't be there.  And again, at some point, someone will want more, and then they will either use `scroll`, or something else, or ask you to implement something very similar, and then what?

Finally, I just really don't have the energy to keep discussing these issues.  I'm not even sure what's been accomplished.  The crate is in a half-implemented state at this point, and it's pretty disheartening.  If you had been pushing to a branch after I gave you write access instead of creating an entirely new crate incrementally without telling me, I'm pretty sure this would be done already and both of us would have what we want.  From my perspective the sequence of events leading up to this point looks somewhat disingenuous to be honest, and isn't the best way to develop open source software.

I believe I've already stated my case for why I don't think you should perform a hard fork/crate which is 2/3rd functionality of scroll, etc.

Please let me know your plans so I can figure out what to do next, thanks.

**Comment by goandylok** [comment]
> As for core ideas, I don't think abandoning generics is smart at all; at least not for looking to future. Once specialization + const lands, I believe will be able to add const generic impls going all the way down to TryFromCtx also taking [u8; 10], etc.

It's ok in specialized solution.
```rust
impl<const N: usize> TryFromCtx for [u8; N]
```
But it doesn't matter, `byte` can be generic and `scroll` can be specialized. That's the same for user.

> With a single crate, I can write data into files, byte slices, vectors, arrays, all generic over the type, which makes refactors easy.

That's not Unix -- do one thing well.

> Removing the offsets from TryFromCtx was an enormous api change, both internally AND externally; it breaks all downstream clients .. this is just what a major refactor looks like

That's why crates should be small, be unix.

>  The crate is in a half-implemented state at this point, and it's pretty disheartening. If you had been pushing to a branch after I gave you write access instead of creating an entirely new crate incrementally without telling me, I'm pretty sure this would be done already and both of us would have what we want. .. and isn't the best way to develop open source software.

Em, in the front-end world, this happens daily. Contribute to improve and open a new library(hard fork) when things goes incompatible. **BUT!! I have never said I stopped the work on `scroll`!!** And actually I am working on that right after I published `byte`. I've done that just because I thought there are naturally incompatibility between my goals and `scroll`(small, panic-free, easy to learn). I certainly want `scroll` to accept all of these, but as you told, it can't be.

I have some options here:

1. Open a pulls to provide returning size to current `scroll`, which is simply copying codes from `byte` and adapt to generic types, and this work is almost done. The pulls can be uploaded these days.

2. Split `scroll` into small pieces. But I don't think this aims your goals.

3. Make `scroll` depend on `byte`, to provide Leb128, io, file operations. In addition, refactor `byte` to use generic slices, to reach a consensus on core design. All these work can be uploaded by pulls on `scroll` by me. And I will open full write access of `byte` to you.

**Comment by goandylok** [comment]
Maybe you can't imagine how serious embedded developer treats panic. Hardware would not abort and give you error message and error exit code, and it wouldn't give you chances to run a debugger after the  disaster. What's worse, if the produce is sold to customers, and unfortunately panicked, you are fired. Hardware has no update opportunity, no feedback method, and maybe expensive or destructive. In my experience, critical numbers are even checked before every addition in C world.

**Comment by goandylok** [comment]
I ran into a problem when trying to implement returning size for lread. I can't know exactly how many bytes are needed before I actually read it.

**Comment by goandylok** [comment]
Maybe we need to give up `lread` at this point.

**Comment by m4b** [comment]
Lread will use FromCtx; I'm ok for now not returning the size in that Trait. You could alternatively comment it out temporarily and I can see what to do

**Comment by m4b** [comment]
@goandylok Weird I didn't get notified for this comment; can you push what you have onto a PR?

I'm eager to get a working version finished as it is now blocking fixing some bugs found while fuzzing goblin.

Not returning size in `FromCtx` is fine, seems ok for now.

**Comment by goandylok** [comment]
The code is half complete but I will be busy the next week. I will restart after that. Sorry for the delay.

**Comment by m4b** [comment]
You can just push what you have its fine, don't care if it compiles right now

**Comment by goandylok** [comment]
I pushed a new branch `develop` with a WIP commit about two weeks ago but it seems not to notify you and my comment is disappeared because of some unknown problems.

My plan was to temporarily depend on `byte` and get `cread` `pread` and `leb128` back, then copy the `byte`'s core code into `scroll`, in generic form.

I picked up the work this week, then I found it difficult to make anything clean when returning size interacted `generic pread`, especially the generic index, which polluted everyone's signatures; it made me quite depressed.

```rust
pub trait TryFromCtx<'a, Ctx = (), This: ?Sized = [u8]>
    where Self: Sized
{
    fn try_from_ctx<I>(src: &'a This, ctx: Ctx) -> Result<(Self, usize), Error<I>>,
        where I: Add + Copy + PartialOrd;
}
```

 I need your opinion about whether to keep on this kind of code style or to give up generic.

**Comment by m4b** [comment]
Omg wth GitHub is pissing me off why I'm not notified. 

Anyway: only briefly checked, but please do not delete files, makes git history strange(r) and if other branches its easier to cause conflicts. Just temporarily comment the module if it's annoying you for some reason

Second, note that the signature for try_from_ctx as provided doesn't make a lot of sense.

But I'm not sure what the problem is TryFromCtx is impl'd for usize and bytes directly, you should be able to index and increment without need for any generics...

Iiuc should only need to change the return signature on the traits and in the macros, and simply return how many bytes it read, or size and it's done, no?

**Comment by m4b** [comment]
So one easy solution is to simply add `type Size` as an associated type to TryFromCtx and then have it be a usize for the impls in the macros, and then return the size appropriately

Shouldn't require much changes at all

**Comment by m4b** [comment]
Epic as fuq.

I would have liked to unify cread and ioread into a single `scroll::Pread` but I don't have enough time to get it working. Alas.
